agent/consul/intention_endpoint.go

```diff
@@ -245,258 +193,212 @@ func (s *Intention) computeApplyChangesLegacyCreate(
-
-	if err := checkLegacyIntentionApplyAllowed(prevEntry); err != nil {
-		return "", nil, err
-	}
-
-	upsertEntry := prevEntry.Clone()
-	upsertEntry.Sources = append(upsertEntry.Sources, args.Intention.ToSourceIntention(true))
 
 	// NOTE: if the append of this source causes a duplicate source name the
 	// config entry validation will fail so we don't have to check that
 	// explicitly here.
 
-	return structs.ConfigEntryUpsertCAS, upsertEntry, nil
+	mut := &structs.IntentionMutation{
+		Destination: args.Intention.DestinationServiceName(),
+		Value:       args.Intention.ToSourceIntention(true),
+	}
+
+	// Set the created/updated times. If this is an update instead of an insert
+	// the UpdateOver() will fix it up appropriately.
+	now := time.Now().UTC()
+	mut.Value.LegacyCreateTime = timePointer(now)
+	mut.Value.LegacyUpdateTime = timePointer(now)
+
+	return mut, nil
 }
 
 func (s *Intention) computeApplyChangesLegacyUpdate(
 	accessorID string,
 	authz acl.Authorizer,
 	entMeta *structs.EnterpriseMeta,
 	args *structs.IntentionRequest,
-) (structs.ConfigEntryOp, *structs.ServiceIntentionsConfigEntry, error) {
+) (*structs.IntentionMutation, error) {
 	// This variant is just for legacy UUID-based intentions.
 
-	_, prevEntry, ixn, err := s.srv.fsm.State().IntentionGet(nil, args.Intention.ID)
+	_, _, ixn, err := s.srv.fsm.State().IntentionGet(nil, args.Intention.ID)
 	if err != nil {
-		return "", nil, fmt.Errorf("Intention lookup failed: %v", err)
-	}
-	if ixn == nil || prevEntry == nil {
-		return "", nil, fmt.Errorf("Cannot modify non-existent intention: '%s'", args.Intention.ID)
+		return nil, fmt.Errorf("Intention lookup failed: %v", err)
 	}
-
-	if err := checkLegacyIntentionApplyAllowed(prevEntry); err != nil {
-		return "", nil, err
+	if ixn == nil {
+		return nil, fmt.Errorf("Cannot modify non-existent intention: '%s'", args.Intention.ID)
 	}
 
-	// Even though the eventual config entry RPC will do an authz check and
-	// validation, if we do them here too we can generate error messages that
-	// make more sense for legacy edits.
 	if !ixn.CanWrite(authz) {
 		// todo(kit) Migrate intention access denial logging over to audit logging when we implement it
 		s.logger.Warn("Update operation on intention denied due to ACLs", "intention", args.Intention.ID, "accessorID", accessorID)
-		return "", nil, acl.ErrPermissionDenied
+		return nil, acl.ErrPermissionDenied
 	}
 
 	args.Intention.DefaultNamespaces(entMeta)
 
 	// Prior to v1.9.0 renames of the destination side of an intention were
 	// allowed, but that behavior doesn't work anymore.
 	if ixn.DestinationServiceName() != args.Intention.DestinationServiceName() {
-		return "", nil, fmt.Errorf("Cannot modify DestinationNS or DestinationName for an intention once it exists.")
+		return nil, fmt.Errorf("Cannot modify DestinationNS or DestinationName for an intention once it exists.")
 	}
 
-	// We always update the updatedat field.
-	args.Intention.UpdatedAt = time.Now().UTC()
```

Comment:
This was already a dead assignment given how `args.Intention.ToSourceIntention` worked.